Network adapter reconfiguration for vCloud provider

miha-plesko opened the pull request:
With this commit we extend reconfigure UI to support what we need for vCloud: addition/deletion of NICs that are connected to cloud networks.
    
Current UI assumed that only vLan (i.e. network) has to be picked when creating new NIC. However, for vCloud we also need to specify new NIC index hence we extend the form a little.
    
BZ: https://bugzilla.redhat.com/show_bug.cgi?id=1572086
Depends on: https://github.com/ManageIQ/manageiq-providers-vmware/pull/272
Depends on: https://github.com/ManageIQ/manageiq-ui-classic/pull/3854

Video: http://x.k00.fr/nicreconfigure

@miq-bot assign @himdel 
@miq-bot add_label enhancement,gaprindashvili/yes,pending core

himdel commented:
(You may need to rebase, looks like the first commit comes from https://github.com/ManageIQ/manageiq-ui-classic/pull/3854.)

miha-plesko commented:
Rebased and introduced two goodies:

- move validation from haml to controller auxilary function
- move checking if `vendor = 'vmware' && type == 'Cloud'` also to controller auxilary function (much more readable)

Please take a look.

himdel commented on app/controllers/mixins/actions/vm_actions/reconfigure.rb:
```diff
@@ -19,7 +19,8 @@ def reconfigure(reconfigure_ids = [])
           if @reconfigitems.size == 1
             vm = Vm.find(@reconfigitems)
-            @vlan_options = get_vlan_options(vm.host_id)
+            @vlan_options = vm.try(:available_networks) || get_vlan_options(vm.host_id)
```
get_vlan_options is only called from this place and returns a list of Lan objects

with this change, you try to call a method available only on one provider and returns list of CloudNetwork objects (not related to Lan)

So - why would you put that try here instead of inside the function written specifically to give you the available lans?

And, why does it make sense to use the same field for 2 unrelated uses?

miha-plesko replied:
Sorry for pissing you off 😃 My thinking was that "Lan" is in fact infra version of "Private Network" for cloud providers, hence I joined them in same field. Didn't reuse `get_vlan_options` function because it relies only on `vm.host_id` which is always `nil` for our cloud provider. But you're right, let me isolate the drop-down from exisiting one completely to improve readability.

himdel replied:
Thanks... I mean, if mixing lans and networks makes sense from the provider perspective, by all means feel free to do so.

I just see Lan & CloudNetwork being mixed in one field and it feel like that's probably a bug :).

miha-plesko replied:
@himdel is it worth investing time in angular drop-down instead rails? If yes, I kindly ask for pointer to example.

himdel replied:
Uh, what would be the difference?

Do you mean using `options_for_select` in the template, versus retrieving the options via the API and populating via `ng-options`?

miha-plesko replied:
That's what I meant, yes, but I'm not sure anymore why would I want to do it this way :)

himdel replied:
Then I would not worry about it too much :)

But if you want.. the idea is that we should be preferring to use the API whenever possible (so that we can remove the UI code which duplicates what the API does).

So.. the select would look like this:
https://github.com/ManageIQ/manageiq-ui-classic/pull/3739/files#diff-31ab4a7ef602ebbddb2e18bde4188f6dR44

And getting the data somewhat like (except not in miqService so you don't have to support custom callbacks :)):
https://github.com/ManageIQ/manageiq-ui-classic/pull/3739/files#diff-e27d4b8817a8b9402efc6a4e6800a565R184


So.. if the API already returns the right thing, deleting the `vlan_options` code is definitely preferred to adding a special case for it :). If not, meh, would be nice.

miha-plesko replied:
I've done as you suggest, thanks for the links. I'm only struggling with empty select value that somehow appears:

![capture](https://user-images.githubusercontent.com/8102426/40619154-47b2440e-6294-11e8-9580-094613b256a4.PNG)

Do you happen to know how to get rig of it?

```haml
              %select{'name'                    => 'adapterNetwork',
                'ng-model'                    => 'vm.reconfigureModel.adapterNetwork',
                'ng-options'                  => 'network for network in vm.reconfigureModel.availableAdapterNetworks',
                'pf-select'                   => true,
                'required'                    => false,
                'selectpicker-for-select-tag' => ''}
              %option{"value" => ""}
                = "<#{_('Choose')}>"
```

himdel commented on app/views/vm_common/_reconfigure.html.haml:
```diff
@@ -343,10 +343,18 @@
           %td
             = select_tag('vLan',
-                       options_for_select([["<#{_('Choose')}>", '']] + @vlan_options, :disabled => ["<#{_('Choose')}>", nil]),
+                       options_for_select([["<#{_('Choose')}>", '']] + @vlan_options),
```
Is chosing no lans is suddenly ok regardless of type?

miha-plesko replied:
No - we make `<Choose>` option selectable in drop-down, but then disable `Confirm Add` button if choosing no lans isn't feasible behavior for provider type (see L367).

himdel replied:
Aah, thanks, makes sense :)

himdel replied:
Actually.. no .. the field is still marked as required

miha-plesko replied:
As far as I've tested adding `"required" => "true"` or `"required" => "false"` makes no difference at all (is ignored for some reason) but no problem, I'll add that.

himdel replied:
Thanks, yeah, the form does not actually use that information, but it really should :)

miha-plesko commented:
I've re-recorded video to use latest code (with angular-driven drop-down): http://x.k00.fr/reconfigurenic
There is a glitch in the dropdown that displays "Nothing Selected" option initially, but I think it's not too big of a problem because it does not affect usability (and because we're in a hurry for 4.6.3 code freeze).

@himdel wdyt?

himdel commented:
@miha-plesko Aah, if you replace `selectpicker-for-select-tag` with `pf-select`, this *should* go away, sorry about that.

himdel commented:
Uh.. actually.. 2 possibilities ... either it's that ^ and you want `pf-select`.
Or you can use either, but need to make sure that the initial ng-model value is an empty string, not undefined or null.

miha-plesko commented:
Dunno, I'm certain that initial value is empty string but still no matter what I do with pf-select and selectpicker-for-select-tag attributes, behavior is the same. Is there perhaps anything else I can improve on this PR other than this tiny issue, or is it ready now? I didn't want to go with `supports` thing here since there is not enough time to merge another core PR.

himdel commented on app/views/vm_common/_reconfigure.html.haml:
```diff
@@ -343,8 +346,25 @@
+                'ng-model'                    => 'vm.reconfigureModel.adapterNetwork',
+                'ng-options'                  => 'network for network in vm.reconfigureModel.availableAdapterNetworks',
+                'pf-select'                   => true,
+                'required'                    => false,
```
Depending on the haml implementation `%x{ :required => false }` is either `<x></x>` or `<x required="false"></x>`. The string `"false"` is still true :).

I suggest you just remove the line.

himdel commented:
I'll try to look into that pf-select issue one of these days, should be trivial to fix in a separate PR.

himdel commented:
Otherwise, no issues from me, waiting for the other PR.

miha-plesko commented:
@himdel green + core PR has been merged

@miq-bot remove-label pending core

simaishi commented:
Gaprindashvili backport details:

```
$ git log -1
commit 74c7514bf0b9433d89222b295be30a7df6bcfb03
Author: Harpreet Kataria <hkataria@redhat.com>
Date:   Wed May 30 15:33:03 2018 -0400

    Merge pull request #3972 from miha-plesko/vcloud-nic-reconfigure
    
    Network adapter reconfiguration for vCloud provider
    (cherry picked from commit 5161e2ed3d7388c12ffdfcca48d9139480622c33)
    
    https://bugzilla.redhat.com/show_bug.cgi?id=1584699
```
